text/0317-third-party-dependency-management.md

```diff
@@ -0,0 +1,129 @@
+like the JSII runtime in case a security vulnerability is found in it.
+Because of that, this RFC only focuses on managing Node dependencies.
+
+### Updating
+
+We have a GitHub Action that once a week updates all dependencies in the CDK repository to their latest,
+usually minor, version,
+using the [`npm-check-updates`](https://www.npmjs.com/package/npm-check-updates) package.
+
+### Summary
+
+Our current situation defends us pretty well against supply-chain attacks,
+because of the bundling that we do
+(our customers are always pinned to a specific version of a dependency).
+
+Also note that CDK is not susceptible to the attack described in the article
+["Researcher hacks over 35 tech firms in novel supply chain attack"](https://www.bleepingcomputer.com/news/security/researcher-hacks-over-35-tech-firms-in-novel-supply-chain-attack),
+as the CDK does not use any non-public packages.
+
+The disadvantage of bundling is that it doesn't make it easy for our customers
+to upgrade any third-party packages that have been found to have security vulnerabilities -
+the customers have to upgrade their CDK package versions in order to get rid of the insecure dependency.
+However, since bundling is a requirement forced on us by JSII,
+there's nothing we can do to change this situation.
+
+## Suggested actions
+
+I suggest we take the following actions:
+
+### 1. Perform a full audit of all third-party dependencies CDK uses
+
+Our goal should be to minimize the number of third-party dependencies we use.
+I suggest we perform an audit of all CDK code,
+and when we determine that a given dependency is only used for a small part of the capabilities it offers,
+like a single function,
+consider vendoring in that functionality into our codebase
+(of course, with the proper license and attribution/copyright)
+in cases when we deem that appropriate.
+
+### 2. Freeze the list of allowed third-party dependencies
+
+After #1 above is completed,
+we should freeze the allowed set of third-party dependencies that we use,
+so that adding a new one will fail by default.
+I have not found an existing tool to perform that sort of locking,
+but I think writing a new `pkglint` rule should be relatively straightforward.
+I imagine the list of allowed third-party dependencies will be kept in a file in source control,
+so we can update it as time goes by,
+and we determine a given new dependency is essential.
+
+### 3. Run a periodic `yarn audit` check
+
+To speed up the feedback cycle between finding out about a
+vulnerability in a version of a third-party dependency we use,
+and fixing it,
+I propose we add a scheduled job that runs `yarn audit` on the CDK repo,
+and creates a ticket in our internal system if any vulnerabilities are found by it.
+
+### 4. Build tooling for deprecating unsafe CDK versions
+
+Because of bundling, once a given third-party dependency version is deemed to have a security vulnerability,
```

Comment:
How will we know this? We will have to periodically run `yarn audit` on all old versions as well, no?

Reply:
Yes. I guess at each release tag?